contrib/db_pools/lib/src/database.rs

```diff
@@ -0,0 +1,137 @@
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl<D: Database> std::ops::DerefMut for Connection<D> {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.0
+    }
+}
+
+#[rocket::async_trait]
+impl<'r, D: Database> rocket::request::FromRequest<'r> for Connection<D> {
+    type Error = <D::Pool as Pool>::GetError;
+
+    async fn from_request(
+        req: &'r rocket::Request<'_>,
+    ) -> rocket::request::Outcome<Self, Self::Error> {
+        use rocket::http::Status;
+        use rocket::request::Outcome;
+
+        let db: &D = match req.rocket().state() {
+            Some(p) => p,
+            _ => panic!(
+                "tried to use database connection pool {}, but its fairing was not attached",
+                D::NAME,
+            ),
+        };
+
+        match db.pool().get().await {
+            Ok(conn) => Outcome::Success(Connection(conn)),
+            Err(e) => Outcome::Failure((Status::ServiceUnavailable, e)),
+        }
+    }
+}
+
+impl<D: Database> Sentinel for Connection<D> {
+    fn abort(rocket: &Rocket<Ignite>) -> bool {
+        use rocket::yansi::Paint;
+
+        if rocket.state::<D>().is_none() {
+            let dbtype = Paint::default(std::any::type_name::<D>()).bold();
+            let fairing = Paint::default(format!("{}::fairing()", dbtype)).wrap().bold();
+            error!("requesting `{}` DB connection without attaching `{}`.", dbtype, fairing);
+            info_!("Attach `{}` to use database connection pooling.", fairing);
+            return true;
+        }
+
+        false
+    }
+}
+
+/// The database fairing for pool types created with the `pool!` macro.
+pub struct Fairing<D: Database>(fn(D::Pool) -> D);
+
+impl<D: Database> Fairing<D> {
+    /// Create a new database fairing with the given constructor.  This
+    /// constructor will be called to create an instance of `D` after the pool
+    /// is initialized and before it is placed into managed state.
+    pub fn new(ctor: fn(D::Pool) -> D) -> Self {
+        Self(ctor)
+    }
+}
+
+#[rocket::async_trait]
+impl<D: Database> rocket::fairing::Fairing for Fairing<D> {
+    fn info(&self) -> Info {
+        Info {
+            name: "rocket_db_pools connection pool",
+            kind: Kind::Ignite,
+        }
+    }
+
+    async fn on_ignite(&self, rocket: Rocket<Build>) -> Result<Rocket<Build>, Rocket<Build>> {
+        let db_config = match rocket
+            .figment()
+            .find_value(&format!("databases.{}", D::NAME))
+        {
+            Ok(v) => v,
+            Err(e) => {
+                error!("error getting database configuration: {}", e);
+                return Err(rocket);
+            }
+        };
+
+        let config = match db_config.deserialize() {
+            Ok(c) => c,
+            Err(e) => {
+                error!("error deserializing configuration: {}", e);
+                return Err(rocket);
+            }
+        };
```

Comment:
This is not how figment should be used here. The database config should be extracted to a pool-specific config structure using figment, not "deserialize". We lose all of the provenance information by doing this. And, write much more code than necessary. Use Figment!

The thing to do here is to:
  1. Create a figment focused on the database config.
  2. Join a set of defaults (size = worker count * k, timeout, etc.).
  3. Extract to the provided config type.

Reply:
Can you describe what changes should be made to the division of work between the trait, impls, and `fairing()` to achieve this? In `sync_db_pools` today, each implementation of `Poolable` must call `Config::from(db_name, rocket)?`. Similarly, I hoped to avoid requiring each implementation of `Pool` (or associated configuration structure) to individually write out e.g. `Figment::from(rocket.figment()).focus(&db_name)` if that boilerplate could be brought into the fairing instead.